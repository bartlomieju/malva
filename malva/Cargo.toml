[package]
name = "malva"
version = "0.0.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "CSS/SCSS/Sass/Less formatter."
repository = "https://github.com/g-plane/malva"
license = "MIT"

[dependencies]
itertools = "0.11"
raffia = "0.4"
tiny_pretty = { path = "../../tiny_pretty" }
