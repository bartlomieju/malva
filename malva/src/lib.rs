mod ctx;
mod doc_gen;
pub mod options;
