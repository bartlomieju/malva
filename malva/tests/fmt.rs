use insta::{assert_snapshot, glob, Settings};
use malva::{config::FormatOptions, format_text};
use raffia::{ast::Stylesheet, Parser, ParserBuilder, Syntax};
use std::fs;

#[test]
fn fmt_snapshot() {
    glob!("fmt/**/*.{css,scss,sass,less}", |path| {
        let input = fs::read_to_string(path).unwrap();
        let syntax = match path.extension().unwrap().to_str().unwrap() {
            "css" => Syntax::Css,
            "scss" => Syntax::Scss,
            "sass" => Syntax::Sass,
            "less" => Syntax::Less,
            _ => unreachable!("unknown file extension"),
        };

        let mut comments = vec![];
        let mut parser = ParserBuilder::new(&input)
            .syntax(syntax)
            .comments(&mut comments)
            .build();
        parser
            .parse::<Stylesheet>()
            .map_err(|error| {
                format!(
                    "failed to parse '{}': {} from {} to {}",
                    path.display(),
                    error.kind,
                    error.span.start,
                    error.span.end
                )
            })
            .unwrap();
        let options = if let Some(config) = comments
            .first()
            .and_then(|comment| comment.content.strip_prefix("cfg"))
        {
            toml::from_str::<FormatOptions>(config).unwrap()
        } else {
            Default::default()
        };

        let output = format_text(&input, syntax, &options).unwrap();
        assert!(
            !output.contains(" \n"),
            "'{}' has trailing whitespace",
            path.display()
        );
        let mut regression_parser = Parser::new(&output, syntax);
        assert!(
            regression_parser.parse::<Stylesheet>().is_ok(),
            "'{}' has syntax errors after formatted",
            path.display()
        );

        let mut settings = Settings::clone_current();
        settings.set_snapshot_path(path.parent().unwrap());
        settings.remove_snapshot_suffix();
        settings.set_prepend_module_to_snapshot(false);
        settings.remove_input_file();
        settings.set_omit_expression(true);
        settings.remove_input_file();
        settings.remove_info();
        settings.bind(|| {
            let name = path.file_stem().unwrap().to_str().unwrap();
            assert_snapshot!(name, output);
        });
    });
}
