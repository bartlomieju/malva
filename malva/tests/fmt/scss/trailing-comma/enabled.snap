---
source: malva/tests/fmt.rs
---
/*cfg trailingComma = true*/
@include section-type-1(
  $header: (margin: 0 0 $margin-base, text-align: left),
  $decoration: (
    type: base,
    margin: 0 auto -1px 0,
    primary-color: $brand-primary,
    secondary-color: $gray-light,
  ),
  $title: (
    margin: 0 0 $margin-small,
    color: false,
    font-size: $font-size-h3,
    font-weight: false,
    line-height: $line-height-h3,
  ),
);

$colors: hotpink deepskyblue firebrick,;
$list: (a,);
$list: ("Helvetica", "Arial", sans-serif);
$colors: ("red", "blue");
$config: (
  themes: (
    mist: (header: #dcfac0, content: #00968b, footer: #85c79c),
    $spring: (header: #f4fac7, content: #c2454e, footer: #ffb158),
  ),
);

$breakpoint-map: (
  small: (
    min-width: null,
    max-width: 479px,
    base-font: 16px,
    vertical-rhythm: 1.3,
  ),
  medium: (
    min-width: 480px,
    max-width: 959px,
    base-font: 18px,
    vertical-rhythm: 1.414,
  ),
  large: (
    min-width: 960px,
    max-width: 1099px,
    base-font: 18px,
    vertical-rhythm: 1.5,
  ),
  xlarge: (
    min-width: 1100px,
    max-width: null,
    base-font: 21px,
    vertical-rhythm: 1.618,
  ),
);
